core/include/traccc/edm/silicon_cell_collection.hpp

```diff
@@ -0,0 +1,203 @@
+    ///
+    TRACCC_HOST_DEVICE
+    auto& module_index() { return BASE::template get<4>(); }
+    /// The index of the module that this cell belongs to (const)
+    ///
+    /// Used to look up the module in @c traccc::silicon_detector_description.
+    ///
+    /// @return A (const) vector of <tt>unsigned int</tt> values
+    ///
+    TRACCC_HOST_DEVICE
+    const auto& module_index() const { return BASE::template get<4>(); }
+
+    /// @}
+
+    /// @name Utility functions
+    /// @{
+
+    /// Assignment operator
+    ///
+    /// @param[in] other The object to assign from
+    /// @return A reference to this object
+    ///
+    TRACCC_HOST_DEVICE silicon_cell& operator=(const silicon_cell& other) {
+        channel0() = other.channel0();
+        channel1() = other.channel1();
+        activation() = other.activation();
+        time() = other.time();
+        module_index() = other.module_index();
+        return *this;
+    }
+
+    /// Assignment operator
+    ///
+    /// @param[in] other The object to assign from
+    /// @return A reference to this object
+    ///
+    template <typename T,
+              std::enable_if_t<!std::is_same_v<BASE, T>, bool> = false>
+    TRACCC_HOST_DEVICE silicon_cell& operator=(const silicon_cell<T>& other) {
+        channel0() = other.channel0();
+        channel1() = other.channel1();
+        activation() = other.activation();
+        time() = other.time();
+        module_index() = other.module_index();
+        return *this;
+    }
+
+    /// Comparison operator
+    ///
+    /// @note This function must only be used on proxy objects, not on
+    ///       containers!
+    ///
+    /// @param[in] other The object to compare with
+    /// @return @c true if the objects are equal, @c false otherwise
+    ///
+    template <typename T>
+    TRACCC_HOST_DEVICE bool operator==(const silicon_cell<T>& other) const {
+        return (channel0() == other.channel0()) &&
+               (channel1() == other.channel1()) &&
+               (math::fabs(activation() - other.activation()) < 1e-6f) &&
+               (math::fabs(time() - other.time()) < 1e-6f) &&
+               (module_index() == other.module_index());
+    }
+
+    /// Ordering operator
+    ///
+    /// @note This function must only be used on proxy objects, not on
+    ///       containers!
+    ///
+    /// @param[in] other The object to compare with
+    /// @return @c true if this object is less than the other, @c false
+    ///         otherwise
+    ///
+    template <typename T>
+    TRACCC_HOST_DEVICE bool operator<(const silicon_cell<T>& other) const {
+        if (module_index() != other.module_index()) {
+            return (module_index() < other.module_index());
+        } else {
+            if (channel1() != other.channel1()) {
+                return (channel1() < other.channel1());
+            } else {
+                return (channel0() < other.channel0());
+            }
+        }
+    }
```

Comment:
These can be unified with an `operator<=>` to reduce boilerplate.

Reply:
That's an interesting idea. :thinking: I'm mainly afraid that these operators are already defined in a bit of a weird way. In order to keep the current beaviour, the spaceship operator would need to be defined in a fairly weird way.

Still, it's an interesting idea. :thinking:

Reply:
Unfortunately the code didn't become much simpler, since `==` is not provided by `<=>`. Still, switching `operator<` to `operator<=>` will make the class a bit more usable.